### ISSUE TITLE
constant: flawed test: hardcoded E2BIG errno value

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#15384]
Migrated from [rt.perl.org#128319](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=128319) (status was 'rejected')

Searchable as RT128319$

**Comment by p5pRT** [comment]
### From @jmdh

As reported in Debian bug \<https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=825018> dist/constant/t/constant\.t has a test which has two separate bugs which conspire to cancel each other out\.

Firstly\, the test hardcodes that E2BIG is supposed to be 7\, but nothing asserts that
in C and POSIX\, and hurd\-i386 happens to be using other values\.

This would cause the test suite to fail\, were it not for a second bug\, which is that the test for the correct error message being produced is rather weak \(it only checks for a long enough string\)&#8203;:

use constant E2BIG => $\! = 7;
cmp\_ok E2BIG\, '=='\, 7;
\# This is something like "Arg list too long"\, but the actual message
\# text may vary\, so we can't test much better than this\.
cmp\_ok length\(E2BIG\)\, '>'\, 6;

I've pushed a patch for the first issue to simply skip the test on Hurd\[1\]\, but it would be nice to have a more accurate test in any case\. Based on observations on different platforms\, would

E2BIG =~ /list too long/i

be a good test\, or are there other variants?

What does 

perl \-e '$\! = 7; print $\! \. "\\n"'

print on other platforms? \(On Linux it's "Argument list too long"\)\.

Note that the reason that this was noticed was that that test \*did\* fail as a side\-effect of \#128279\, now also fixed\.\.\.

\[1\] \<http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/21866d1330b560599ac45bc24574bbd830e59a37>

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Jun 04 07&#8203;:01&#8203;:08 2016\, dom wrote&#8203;:
> As reported in Debian bug \<https://bugs.debian.org/cgi-
> bin/bugreport\.cgi?bug=825018> dist/constant/t/constant\.t has a test
> which has two separate bugs which conspire to cancel each other out\.
> 
> Firstly\, the test hardcodes that E2BIG is supposed to be 7\, but
> nothing asserts that
> in C and POSIX\, and hurd\-i386 happens to be using other values\.
> 
> This would cause the test suite to fail\,

I do not see how\.  E2BIG here is the name of a constant created by constant\.pm and then immediately tested to see whether its value is set as expected\.  Presumably the test is supposed to mimic a real\-world example\.  But the constant could have been named EWAHATVEER as far as I can tell\.

> were it not for a second bug\,
> which is that the test for the correct error message being produced is
> rather weak \(it only checks for a long enough string\)&#8203;:
> 
> use constant E2BIG => $\! = 7;
> cmp\_ok E2BIG\, '=='\, 7;
> \# This is something like "Arg list too long"\, but the actual message
> \# text may vary\, so we can't test much better than this\.
> cmp\_ok length\(E2BIG\)\, '>'\, 6;
> 
> I've pushed a patch for the first issue to simply skip the test on
> Hurd\[1\]\, but it would be nice to have a more accurate test in any
> case\. Based on observations on different platforms\, would
> 
> E2BIG =~ /list too long/i
> 
> be a good test\, or are there other variants?

How about testing that E2BIG eq "$\!"?  The whole point is to see whether use constant FOO => $\!=\.\.\. will create a dualvar constant\.

> Note that the reason that this was noticed was that that test \*did\*
> fail as a side\-effect of \#128279\, now also fixed\.\.\.

Was is the cmp\_ok E2BIG\, '=='\, 7 that failed? or the length test?  If the stringification of $\! was broken\, then the second test would have failed\, and the fact that E2BIG is not 7 on Hurd should be irrelevant as long as $\! stringifies properly\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @geeknik

On Sat Jun 04 07&#8203;:01&#8203;:08 2016\, dom wrote&#8203;:
> What does
> 
> perl \-e '$\! = 7; print $\! \. "\\n"'
> 
> print on other platforms? \(On Linux it's "Argument list too long"\)\.

perl on FreeBSD 10\.3\-RELEASE \#6 returns "Argument list too long" as well\.

**Comment by p5pRT** [comment]
### From @jmdh

On Sat\, Jun 04\, 2016 at 10&#8203;:09&#8203;:31AM \-0700\, Father Chrysostomos via RT wrote&#8203;:
> On Sat Jun 04 07&#8203;:01&#8203;:08 2016\, dom wrote&#8203;:
> > As reported in Debian bug \<https://bugs.debian.org/cgi-
> > bin/bugreport\.cgi?bug=825018> dist/constant/t/constant\.t has a test
> > which has two separate bugs which conspire to cancel each other out\.
> > 
> > Firstly\, the test hardcodes that E2BIG is supposed to be 7\, but
> > nothing asserts that
> > in C and POSIX\, and hurd\-i386 happens to be using other values\.
> > 
> > This would cause the test suite to fail\,
> 
> I do not see how\.  E2BIG here is the name of a constant created by constant\.pm and then immediately tested to see whether its value is set as expected\.  Presumably the test is supposed to mimic a real\-world example\.  But the constant could have been named EWAHATVEER as far as I can tell\.

Clumsy wording\. The point is that the test is trying to check that
7 should stringify to something along the lines of "Arg list too long"
but that isn't the case on Hurd&#8203;:

dom@&#8203;exodar&#8203;:\~$ perl \-e '$\! = 7; print $\! \. "\\n"'
\(os/kern\) not receiver

> > were it not for a second bug\,
> > which is that the test for the correct error message being produced is
> > rather weak \(it only checks for a long enough string\)&#8203;:
> > 
> > use constant E2BIG => $\! = 7;
> > cmp\_ok E2BIG\, '=='\, 7;
> > \# This is something like "Arg list too long"\, but the actual message
> > \# text may vary\, so we can't test much better than this\.
> > cmp\_ok length\(E2BIG\)\, '>'\, 6;
> > 
> > I've pushed a patch for the first issue to simply skip the test on
> > Hurd\[1\]\, but it would be nice to have a more accurate test in any
> > case\. Based on observations on different platforms\, would
> > 
> > E2BIG =~ /list too long/i
> > 
> > be a good test\, or are there other variants?
> 
> How about testing that E2BIG eq "$\!"?  The whole point is to see whether use constant FOO => $\!=\.\.\. will create a dualvar constant\.
> 
> > Note that the reason that this was noticed was that that test \*did\*
> > fail as a side\-effect of \#128279\, now also fixed\.\.\.
> 
> Was is the cmp\_ok E2BIG\, '=='\, 7 that failed? or the length test?  If the stringification of $\! was broken\, then the second test would have failed\, and the fact that E2BIG is not 7 on Hurd should be irrelevant as long as $\! stringifies properly\.

The latter\. I'm only mentioning this because it was what caused the
issue to be noticed\, not because it's fundamental to \*this\* bug\.

It's probably easier to read the whole thread on
\<https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=825018> as my attempts
to summarise seem to have caused more confusion\.

Dominic\.

**Comment by p5pRT** [comment]
### From @khwilliamson

On 06/04/2016 12&#8203;:44 PM\, Dominic Hargreaves wrote&#8203;:
> On Sat\, Jun 04\, 2016 at 10&#8203;:09&#8203;:31AM \-0700\, Father Chrysostomos via RT wrote&#8203;:
>> On Sat Jun 04 07&#8203;:01&#8203;:08 2016\, dom wrote&#8203;:
>>> As reported in Debian bug \<https://bugs.debian.org/cgi-
>>> bin/bugreport\.cgi?bug=825018> dist/constant/t/constant\.t has a test
>>> which has two separate bugs which conspire to cancel each other out\.
>>>
>>> Firstly\, the test hardcodes that E2BIG is supposed to be 7\, but
>>> nothing asserts that
>>> in C and POSIX\, and hurd\-i386 happens to be using other values\.
>>>
>>> This would cause the test suite to fail\,
>>
>> I do not see how\.  E2BIG here is the name of a constant created by constant\.pm and then immediately tested to see whether its value is set as expected\.  Presumably the test is supposed to mimic a real\-world example\.  But the constant could have been named EWAHATVEER as far as I can tell\.
>
> Clumsy wording\. The point is that the test is trying to check that
> 7 should stringify to something along the lines of "Arg list too long"
> but that isn't the case on Hurd&#8203;:
>
> dom@&#8203;exodar&#8203;:\~$ perl \-e '$\! = 7; print $\! \. "\\n"'
> \(os/kern\) not receiver
>
>>> were it not for a second bug\,
>>> which is that the test for the correct error message being produced is
>>> rather weak \(it only checks for a long enough string\)&#8203;:
>>>
>>> use constant E2BIG => $\! = 7;
>>> cmp\_ok E2BIG\, '=='\, 7;
>>> \# This is something like "Arg list too long"\, but the actual message
>>> \# text may vary\, so we can't test much better than this\.
>>> cmp\_ok length\(E2BIG\)\, '>'\, 6;
>>>
>>> I've pushed a patch for the first issue to simply skip the test on
>>> Hurd\[1\]\, but it would be nice to have a more accurate test in any
>>> case\. Based on observations on different platforms\, would
>>>
>>> E2BIG =~ /list too long/i
>>>
>>> be a good test\, or are there other variants?
>>
>> How about testing that E2BIG eq "$\!"?  The whole point is to see whether use constant FOO => $\!=\.\.\. will create a dualvar constant\.
>>
>>> Note that the reason that this was noticed was that that test \*did\*
>>> fail as a side\-effect of \#128279\, now also fixed\.\.\.
>>
>> Was is the cmp\_ok E2BIG\, '=='\, 7 that failed? or the length test?  If the stringification of $\! was broken\, then the second test would have failed\, and the fact that E2BIG is not 7 on Hurd should be irrelevant as long as $\! stringifies properly\.
>
> The latter\. I'm only mentioning this because it was what caused the
> issue to be noticed\, not because it's fundamental to \*this\* bug\.
>
> It's probably easier to read the whole thread on
> \<https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=825018> as my attempts
> to summarise seem to have caused more confusion\.
>
> Dominic\.
>


The correct value for E2BIG should be available via POSIX\.  Why not use 
that?

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Jun 04 11&#8203;:45&#8203;:12 2016\, dom wrote&#8203;:
> On Sat\, Jun 04\, 2016 at 10&#8203;:09&#8203;:31AM \-0700\, Father Chrysostomos via RT
> wrote&#8203;:
> > On Sat Jun 04 07&#8203;:01&#8203;:08 2016\, dom wrote&#8203;:
> > > As reported in Debian bug \<https://bugs.debian.org/cgi-
> > > bin/bugreport\.cgi?bug=825018> dist/constant/t/constant\.t has a test
> > > which has two separate bugs which conspire to cancel each other
> > > out\.
> > >
> > > Firstly\, the test hardcodes that E2BIG is supposed to be 7\, but
> > > nothing asserts that
> > > in C and POSIX\, and hurd\-i386 happens to be using other values\.
> > >
> > > This would cause the test suite to fail\,
> >
> > I do not see how\.  E2BIG here is the name of a constant created by
> > constant\.pm and then immediately tested to see whether its value is
> > set as expected\.  Presumably the test is supposed to mimic a real\-
> > world example\.  But the constant could have been named EWAHATVEER as
> > far as I can tell\.
> 
> Clumsy wording\. The point is that the test is trying to check that
> 7 should stringify to something along the lines of "Arg list too long"

Actually it’s not testing that \(even if that may have been the original intent when the test was first added\)\.

> but that isn't the case on Hurd&#8203;:
> 
> dom@&#8203;exodar&#8203;:\~$ perl \-e '$\! = 7; print $\! \. "\\n"'
> \(os/kern\) not receiver
> 
> > > were it not for a second bug\,
> > > which is that the test for the correct error message being produced
> > > is
> > > rather weak \(it only checks for a long enough string\)&#8203;:

Whether the correct error messages is produced is not relevant to testing whether constant can correctly produce a dualvar from $\!\.

> > >
> > > use constant E2BIG => $\! = 7;
> > > cmp\_ok E2BIG\, '=='\, 7;
> > > \# This is something like "Arg list too long"\, but the actual
> > > message
> > > \# text may vary\, so we can't test much better than this\.
> > > cmp\_ok length\(E2BIG\)\, '>'\, 6;
> > >
> > > I've pushed a patch for the first issue to simply skip the test on
> > > Hurd\[1\]\, but it would be nice to have a more accurate test in any
> > > case\. Based on observations on different platforms\, would
> > >
> > > E2BIG =~ /list too long/i
> > >
> > > be a good test\, or are there other variants?
> >
> > How about testing that E2BIG eq "$\!"?  The whole point is to see
> > whether use constant FOO => $\!=\.\.\. will create a dualvar constant\.
> >
> > > Note that the reason that this was noticed was that that test \*did\*
> > > fail as a side\-effect of \#128279\, now also fixed\.\.\.
> >
> > Was is the cmp\_ok E2BIG\, '=='\, 7 that failed? or the length test?  If
> > the stringification of $\! was broken\, then the second test would have
> > failed\, and the fact that E2BIG is not 7 on Hurd should be irrelevant
> > as long as $\! stringifies properly\.
> 
> The latter\. I'm only mentioning this because it was what caused the
> issue to be noticed\, not because it's fundamental to \*this\* bug\.
> 
> It's probably easier to read the whole thread on
> \<https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=825018> as my
> attempts
> to summarise seem to have caused more confusion\.

I have just read it\.  With all due respect\, I think you are missing the forest for the trees\.  I see nothing wrong with the test\, but perhaps the comments surrounding it could use some clarification\.

If I were to write a test like this&#8203;:

\# Test that constant\.pm can create a dualvar out of $\!
use constant A\_DUALVAR\_CONSTANT => $\! = 7;
cmp\_ok A\_DUALVAR\_CONSTANT\, '=='\, 7;
\# Make sure we have an error message string\.  It does not
\# matter that 7 means different things on different platforms\.
\# If this test fails\, then either constant\.pm or $\! is broken&#8203;:
cmp\_ok length\(A\_DUALVAR\_CONSTANT\)\, '>'\, 6;

I hope you would have no reason to skip it on any platform\.  That is functionally identical to what constant\.t does\.

The only reason it failed on Hurd was that $\! was broken\, so it was correct for it to fail\.

Skipping the tests on Hurd\, which you have done\, is not the correct thing to do\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Jun 04 12&#8203;:28&#8203;:36 2016\, public@&#8203;khwilliamson\.com wrote&#8203;:
> The correct value for E2BIG should be available via POSIX\.  Why not
> use
> that?

If we were to do that\, the test should go in t/op/magic\.t\, since it would be testing $\! behaviour\.  constant\.pm does not need to be involved in such a test\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @jmdh

On Sat Jun 04 16&#8203;:26&#8203;:40 2016\, sprout wrote&#8203;:
> On Sat Jun 04 11&#8203;:45&#8203;:12 2016\, dom wrote&#8203;:
> > On Sat\, Jun 04\, 2016 at 10&#8203;:09&#8203;:31AM \-0700\, Father Chrysostomos via RT
> > wrote&#8203;:
> > > On Sat Jun 04 07&#8203;:01&#8203;:08 2016\, dom wrote&#8203;:
> > > > As reported in Debian bug \<https://bugs.debian.org/cgi-
> > > > bin/bugreport\.cgi?bug=825018> dist/constant/t/constant\.t has a
> > > > test
> > > > which has two separate bugs which conspire to cancel each other
> > > > out\.
> > > >
> > > > Firstly\, the test hardcodes that E2BIG is supposed to be 7\, but
> > > > nothing asserts that
> > > > in C and POSIX\, and hurd\-i386 happens to be using other values\.
> > > >
> > > > This would cause the test suite to fail\,
> > >
> > > I do not see how\.  E2BIG here is the name of a constant created by
> > > constant\.pm and then immediately tested to see whether its value is
> > > set as expected\.  Presumably the test is supposed to mimic a real\-
> > > world example\.  But the constant could have been named EWAHATVEER
> > > as
> > > far as I can tell\.
> >
> > Clumsy wording\. The point is that the test is trying to check that
> > 7 should stringify to something along the lines of "Arg list too
> > long"
> 
> Actually it’s not testing that \(even if that may have been the
> original intent when the test was first added\)\.
> 
> > but that isn't the case on Hurd&#8203;:
> >
> > dom@&#8203;exodar&#8203;:\~$ perl \-e '$\! = 7; print $\! \. "\\n"'
> > \(os/kern\) not receiver
> >
> > > > were it not for a second bug\,
> > > > which is that the test for the correct error message being
> > > > produced
> > > > is
> > > > rather weak \(it only checks for a long enough string\)&#8203;:
> 
> Whether the correct error messages is produced is not relevant to
> testing whether constant can correctly produce a dualvar from $\!\.
> 
> > > >
> > > > use constant E2BIG => $\! = 7;
> > > > cmp\_ok E2BIG\, '=='\, 7;
> > > > \# This is something like "Arg list too long"\, but the actual
> > > > message
> > > > \# text may vary\, so we can't test much better than this\.
> > > > cmp\_ok length\(E2BIG\)\, '>'\, 6;
> > > >
> > > > I've pushed a patch for the first issue to simply skip the test
> > > > on
> > > > Hurd\[1\]\, but it would be nice to have a more accurate test in any
> > > > case\. Based on observations on different platforms\, would
> > > >
> > > > E2BIG =~ /list too long/i
> > > >
> > > > be a good test\, or are there other variants?
> > >
> > > How about testing that E2BIG eq "$\!"?  The whole point is to see
> > > whether use constant FOO => $\!=\.\.\. will create a dualvar constant\.
> > >
> > > > Note that the reason that this was noticed was that that test
> > > > \*did\*
> > > > fail as a side\-effect of \#128279\, now also fixed\.\.\.
> > >
> > > Was is the cmp\_ok E2BIG\, '=='\, 7 that failed? or the length test?
> > > If
> > > the stringification of $\! was broken\, then the second test would
> > > have
> > > failed\, and the fact that E2BIG is not 7 on Hurd should be
> > > irrelevant
> > > as long as $\! stringifies properly\.
> >
> > The latter\. I'm only mentioning this because it was what caused the
> > issue to be noticed\, not because it's fundamental to \*this\* bug\.
> >
> > It's probably easier to read the whole thread on
> > \<https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=825018> as my
> > attempts
> > to summarise seem to have caused more confusion\.
> 
> I have just read it\.  With all due respect\, I think you are missing
> the forest for the trees\.  I see nothing wrong with the test\, but
> perhaps the comments surrounding it could use some clarification\.
> 
> If I were to write a test like this&#8203;:
> 
> \# Test that constant\.pm can create a dualvar out of $\!
> use constant A\_DUALVAR\_CONSTANT => $\! = 7;
> cmp\_ok A\_DUALVAR\_CONSTANT\, '=='\, 7;
> \# Make sure we have an error message string\.  It does not
> \# matter that 7 means different things on different platforms\.
> \# If this test fails\, then either constant\.pm or $\! is broken&#8203;:
> cmp\_ok length\(A\_DUALVAR\_CONSTANT\)\, '>'\, 6;
> 
> I hope you would have no reason to skip it on any platform\.  That is
> functionally identical to what constant\.t does\.
> 
> The only reason it failed on Hurd was that $\! was broken\, so it was
> correct for it to fail\.
> 
> Skipping the tests on Hurd\, which you have done\, is not the correct
> thing to do\.

Thanks for the explanation\. I have pushed a change re\-enabling tests on Hurd and improving the test comments along the lines you've suggested\.

Dominic\.

**Comment by p5pRT** [comment]
@jmdh - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Jun 05 03&#8203;:57&#8203;:19 2016\, dom wrote&#8203;:
> Thanks for the explanation\. I have pushed a change re\-enabling tests
> on Hurd and improving the test comments along the lines you've
> suggested\.
> 
> Dominic\.

Thank you for being persuaded\. :\-\)

\-\- 

Father Chrysostomos
